### PR TITLE
Implement binary format for test outputs and errors

### PR DESCRIPTION
**Pull request by kolumb**
Tried to implement binary format that was described in #50

### CODE REVIEW

**Comment by drocha87** [comment]
@kolumb since you are working on the tests I would like to point something. There are some tests which have a long output (like `stack.porth` for example), and then is a little bit hard to spot where the error is (imagine a more complex output).

I don't know if is possible or simple to implement a way to tell where the output has diverged, something like running a `diff` (if the user has it installed) and then printing the result to output, or something like this. 

Maybe one way to solve this is breaking the tests in small pieces and have more detailed tests somehow.

**Comment by drocha87** [comment]
And I think the `test.py` should have a flag to stop the tests as soon as it fails or maybe a flag to run the tests silently and only output the broken tests, what you think?

**Comment by kolumb** [comment]
I think that it could be reported as separate issue that this PR introduces.
Because, if I understand correctly, right now main priority that author has is to rewrite compiler and test script in porth. So I tried to keep everything as simple as possible. I contributed only because I thought that this PR could be helpful for rewriting process.

File: README.md [hunk 1]
```diff
@@ -64,7 +64,7 @@ $ ./program
 
 ### Testing
 
-Test cases are located in [./tests/](./tests/) folder. The `*.txt` files are the expected outputs of the corresponding programs.
+Test cases are located in [./tests/](./tests/) folder. The `*.bin` files are the expected outputs of the corresponding programs in binary format (1 byte - return code, 8 bytes - length of output, output in utf-8 format, 8 bytes - length of error, error in utf-8 format).
```
Comment (rexim) [inline README.md:67]: Please present the format as it was described originally inside of a code block:
```
|1 byte -- expected return code|
|8 bytes -- length of stdout|
|len(stdout) bytes -- the expected stdout encoded as UTF-8|
|8 bytes -- length of stderr|
|len(stderr) bytes -- the expected stderr encoded as UTF-8|
```

File: test.py [hunk 2]
```diff
@@ -22,30 +17,52 @@ def test(folder):
...
             print('[INFO] Testing %s' % entry.path)
 
-            txt_path = entry.path[:-len(porth_ext)] + ".txt"
+            bin_path = entry.path[:-len(porth_ext)] + ".bin"
             expected_output = None
-            with open(txt_path, "rb") as f:
-                expected_output = f.read()
+            with open(bin_path, "rb") as f:
+                bin_file = f.read()
+                index = 0
```
Comment (rexim) [inline test.py:24]: The `read` method has an optional `size` in bytes. You can read the chunks of corresponding size without maintaining this `index`.

File: test.py [hunk 3]
```diff
@@ -55,18 +72,23 @@ def record(folder, mode='sim'):
...
                 print("[ERROR] Unknown record mode `%s`" % mode)
                 exit(1)
-            txt_path = entry.path[:-len(porth_ext)] + ".txt"
-            print("[INFO] Saving output to %s" % txt_path)
-            with open(txt_path, "wb") as txt_file:
-                txt_file.write(output)
+            bin_path = entry.path[:-len(porth_ext)] + ".bin"
+            print("[INFO] Saving output to %s" % bin_path)
+            with open(bin_path, "wb") as bin_file:
+                bin_file.write(
```
Comment (rexim) [inline test.py:85]: Split this into several writes. No need to concatenate anything.

**Review by rexim — CHANGES_REQUESTED** [review]

**Comment by rexim** [comment]
The fact that binary files are not displayed properly in diffs is definitely an issue. I'll need to think about a better simple format.

**Comment by rexim** [comment]
Something like this. What do you guys think?

**Review by rexim — APPROVED** [review]

**Comment by rexim** [comment]
@kolumb thank you for the contribution! :+1:

**Pull request merged**